feat(optimization): Add purity analysis to SSA

jfecher opened the pull request:
# Description

## Problem\*

Resolves https://github.com/noir-lang/noir/issues/7196

## Summary\*

Adds a purity analysis pass to approximate the purity of each function (we only fail for mutually recursive functions) and use its results in other SSA passes.

## Additional Context



## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

jfecher commented:
Pending some tests but the pass itself is finished

TomAFrench commented on compiler/noirc_evaluator/src/ssa/opt/pure.rs:
```diff
@@ -0,0 +1,269 @@
+                            },
+                            // The function we're calling is unknown in the remaining cases,
+                            // so just assume the worst.
+                            Value::ForeignFunction(_)
```
Nit but I think it's worth separating out `ForeignFunction` to call out that foreign calls are 100% impure whereas the other cases are unreachable off of the top of my head.

jfecher replied:
Done

TomAFrench commented on compiler/noirc_evaluator/src/ssa/opt/pure.rs:
```diff
@@ -0,0 +1,269 @@
+    finished_purities
+}
+
+fn build_call_graph(
+    dependencies: HashMap<FunctionId, BTreeSet<FunctionId>>,
+) -> (DiGraph<FunctionId, ()>, HashMap<FunctionId, PetGraphIndex>, HashMap<PetGraphIndex, FunctionId>)
+{
+    let mut graph = DiGraph::new();
+    let mut ids_to_indices = HashMap::default();
+    let mut indices_to_ids = HashMap::default();
+
+    for function in dependencies.keys() {
+        let index = graph.add_node(*function);
+        ids_to_indices.insert(*function, index);
+        indices_to_ids.insert(index, *function);
+    }
+
+    // Create edges from caller -> called
+    for (function, dependencies) in dependencies {
+        let function_index = ids_to_indices[&function];
+
+        for dependency in dependencies {
+            let dependency_index = ids_to_indices[&dependency];
+            graph.add_edge(function_index, dependency_index, ());
+        }
+    }
+
+    (graph, ids_to_indices, indices_to_ids)
+}
```
We have some code in the inlining pass which builds up a callgraph so there's potential for code reuse here.

jfecher replied:
I see compute_bottom_up_order but I'll leave them as separate for now since there's enough logic there having to do with weights that I think they are reasonably different. We probably should have a pass agnostic function to create a call graph eventually though. Perhaps this one can be reorganized in a later refactor.

Another thing I ran into was that I tried to account for mutually recursive functions in my first approach but would have needed a call graph which grouped SCCs into the same node which proved difficult. Having a call graph variant that did that (more expensive) as well would be nice.

vezenovm replied:
I'm also building a call graph in https://github.com/noir-lang/noir/pull/7188, but with slightly different data than in the inlining pass and this pass. I think it would be good to keep each PR as they are and then refactor later to create a pass agnostic call graph.

jfecher commented:
Moved the purity pass, and added another closer to the end of SSA.
I've found the results can change as different calls get optimized out so I opted to move the call closer to:
1. The loop invariant code motion pass
2. The constant folding pass

jfecher commented:
This PR still doesn't seem to be deduplicating
```
fn main(x: Field) {
    unsafe {
        let x1 = foo(x);
        let x2 = foo(x);
        assert_eq(x1, 0);
        assert_eq(x2, 0);
    }
}

unconstrained fn foo(x: Field) -> Field {
    x
}
```

TomAFrench commented:
We don't currently do any deduplication of brillig calls so I think we should add a test to ensure that we don't end up deduplicating a brillig call such that an incorrect predicate is applied.

```nr
fn main(x: Field, active: bool) {
    let mut x1 = false;
    unsafe {
        if active {
            x1 = foo(x);        
        }
    
        let x2 = foo(x);
        assert_eq(x2);
    }
}

unconstrained fn foo(x: Field) -> bool {
    true
}
```

jfecher commented:
I've also found the reason we weren't deduplicating was that we've been introducing different ValueIds for the same function:
```
acir(inline) predicate_pure fn main f0 {
  b0(v0: Field):
    v2 = call (v1) f1(v0) -> Field
    v4 = call (v3) f1(v0) -> Field
    constrain v2 == Field 0
    constrain v4 == Field 0
    return
}
```

jfecher commented:
Had to rewrite a good chunk of the defunctionalize pass because it depended on ValueIds for Value::Function _not_ being unique so that it could map value ids for higher order functions while keeping value ids that were called directly. PR should be ready for review now though it's a bit heftier.

vezenovm commented:
> Looks like (I'm guessing) my defunctionalize changes come with a hefty compile-time setback though I've yet to test what specifically it is from.

Yeah looking at the size of the regressions it would be good to discern where exactly it is coming form. Perhaps we should pull out the defunctionalization rework into a separate PR.

jfecher commented:
I think the performance alert here may be bugged. I tested sha256_regression locally and found that this PR compiles  it slightly faster, not several times slower. I also opened https://github.com/noir-lang/noir/pull/7227 to run the CI on a fresh PR and it shows no performance regression comment there.

TomAFrench commented:
I think it's basically correct but not the best at flagging that a regression has been resolved.

jfecher commented:
Since there are no perf regressions this is ready to review again at least

TomAFrench reviewed (approved):
LGTM. Only followup improvement (which I expect won't catch much that this PR doesn't catch already)

jfecher commented on compiler/noirc_evaluator/src/ssa/opt/pure.rs:
```diff
@@ -0,0 +1,367 @@
+        let mut result = if self.runtime().is_acir() {
+            Purity::Pure
+        } else {
+            // Because we return bogus values when a brillig function is called from acir
+            // in a disabled predicate, brillig functions can never be truly pure unfortunately.
+            Purity::PureWithPredicate
```
Made an issue here: https://github.com/noir-lang/noir/issues/7229
